### PR TITLE
scripts: twisterlib: coverage: fix multiple branch excludes

### PR DESCRIPTION
**Pull request by JordanYates**
Multiple values for `--exclude-branches-by-pattern` will result in only
the last value taking effect. Resolve this by merging all the provided
regex patterns into a single pattern with the `|` operator.

Disable branch coverage for the `__ASSERT` family of macros. Covering
all of the assertion branches by definition means triggering the
assertion, which can be either challenging or impossible to exercise,
and in either case results in the immediate termination of the test.

### CODE REVIEW

**Comment by JordanYates** [comment]
If disabling coverage of `__ASSERT` counts as a feature for the purposes of the release, I can drop the commit.

File: scripts/pylib/twister/twisterlib/coverage.py [hunk 1]
```diff
@@ -330,7 +330,10 @@ def _generate(self, outdir, coveragelog):
         ztestfile = os.path.join(outdir, "ztest.json")
 
         excludes = Gcovr._interleave_list("-e", self.ignores)
-        excludes += Gcovr._interleave_list("--exclude-branches-by-pattern", self.ignore_branch_patterns)
+        if len(self.ignore_branch_patterns) > 0:
```
Comment (LukaszMrugala) [inline scripts/pylib/twister/twisterlib/coverage.py:333]: Just `if self.ignore_branch_patterns:` would suffice and cover the case where this is `None`, but that situation should, in theory, never happen. Can stay as-is.

File: scripts/pylib/twister/twisterlib/coverage.py [hunk 2]
```diff
@@ -425,5 +428,8 @@ def run_coverage(testplan, options):
     # Ignore branch coverage on LOG_* and LOG_HEXDUMP_* macros
     # Branch misses are due to the implementation of Z_LOG2 and cannot be avoided
     coverage_tool.add_ignore_branch_pattern(r"^\s*LOG_(?:HEXDUMP_)?(?:DBG|INF|WRN|ERR)\(.*")
+    # Ignore branch coverage on __ASSERT* macros
+    # Covering the failing case is not desirable as it will immediately terminate the test.
+    coverage_tool.add_ignore_branch_pattern(r"^\s*__ASSERT(?:_EVAL|_NO_MSG|_POST_ACTION)?\(.*")
```
Comment (LukaszMrugala) [inline scripts/pylib/twister/twisterlib/coverage.py:433]: This change seems so separate from the first that it should, in my opinion, be a separate PR.

**Review by LukaszMrugala — CHANGES_REQUESTED** [review]

File: scripts/pylib/twister/twisterlib/coverage.py (hunk 1 above)
Comment (JordanYates) [reply scripts/pylib/twister/twisterlib/coverage.py:333]: It would suffice, but it is less informative regarding the expected type and is less useful IMO than `if len(array):`, which itself is picked up by various tools as being unclear

**Review by LukaszMrugala — APPROVED** [review]

**Review by nashif — APPROVED** [review]
